Significant location changes / deferred location updates

MatthewGerber opened the pull request:
Fixes #222.

Changes Proposed in this pull request:
- Add ability to monitor significant location changes in iOS.
- Add ability to defer location updates in iOS.

This fixes/implements:
- [ ] Bug
- [x] Feature Request

Which plugin does this impact:
- [ ] Battery
- [ ] Connectivity
- [ ] Contacts
- [ ] DeviceInfo
- [ ] ExternalMaps
- [x] Geolocator
- [ ] Media
- [ ] Permissions
- [ ] Settings
- [ ] Text To Speech
- [ ] Vibrate
- Other:

MatthewGerber commented:
I haven't yet added significant location change monitoring, but I've added deferred location updates.

Am I on the right track given your coding approach?

MatthewGerber commented:
Also, I'm on a Mac and can't load the UWP project, so the build checks are failing.

jamesmontemagno commented:
So this is a tricky one and here are my thoughts:

## DeferralSettings
I think that instead of expanding the parameters ( bool defersLocationUpdates = false, double deferralDistanceMeters = -1, double deferralTimeSeconds = -1)

Let's make this a class called Deferral Settings that you have to pass in and it defaults to null:

```csharp
public class DeferLocationUpdateSettings
{
    public double DistanceInMeters {get;set;} = 100; // some default that is good
    public TimeSpan TimeBetweenUpdates {get; get;}  = TimeSpan.FromMintues(10); // or some default idk
}
```

I don't really think this is a setting that will be changed at all to be honest with you. I would remove the 

```
public bool DefersLocationUpdates { get; set; }
```

and replace it with just a:
DeferLocationUpdateSettings deferSettings;

We would set this when they pass it in.

And if they pass in the setting then we would enable it and then when they StopListening() we would turn it off and null it out.


Also reading the docs it says:
This method requires GPS hardware to be available, CLLocationManager.DistanceFilter to be CLLocationDistance.None, and CLLocationManager.DesiredAccuracy be either CLLocation.AccuracyBest or CLLocation.AccuracyBestForNavigation. 

We should just set that to be sure.


For UWP you will have to open up the source code file and modify the constructor after these changes to get it building in Appveyor

MatthewGerber commented:
@jamesmontemagno I think I've successfully added both significant change listening and deferral of location updates. Can you take a look and publish a new Nuget if it looks good to you?

jamesmontemagno commented:
@MatthewGerber I am getting ready to leave for a few days of holiday, but will take a peak and will release as a beta nuget first

jamesmontemagno commented on Geolocator/Geolocator/Geolocator.Plugin.Abstractions/EnergySettings.cs:
```diff
@@ -0,0 +1,34 @@
+		/// Gets or sets a value indicating whether the location manager should defer location updates until an energy efficient time arrives, or distance and time criteria are met. Only available on iOS 6.0 and above.
+		/// </summary>
+		/// <value><c>true</c> to defer location updates; otherwise, <c>false</c> (default).</value>
+		public bool DeferLocationUpdates { get; set; } = false;
```
I am a bit confused between this and ListenForSignificantChanges... They seem to be the same thing.

MatthewGerber replied:
My understanding is that there are two modes of location sensing:  (1) the standard service, which is used in typical GPS-based mapping. This is a continuous monitoring mode that uses significant power. (2) The significant change service, which only reports large changes in locations and consumes less power. In the above code, `ListenForSignificantChanges` selects between these two modes.

iOS supports deferral of location updates for either of the two modes above. If deferral is not used, then any location updates (from either GPS mode above) are delivered immediately to the app, waking the app up in the case of significant changes. If deferral is used, all updates are cached until a delivery criterion (based on distance or time) is met. When this happens, all updates are delivered. In the code above, `DeferLocationUpdates` selects the deferral mode (on/off).

jamesmontemagno commented on Geolocator/Geolocator/Geolocator.Plugin.Abstractions/EnergySettings.cs:
```diff
@@ -0,0 +1,34 @@
+		/// If deferring location updates, the minimum distance to travel before updates are delivered. Set to null for infinite wait.
+		/// </summary>
+		/// <value>The deferral distance meters (default:  500).</value>
+		public double? DeferralDistanceMeters { get; set; } = 500;
```
Any specific reason for nullables?

MatthewGerber replied:
iOS supports unlimited distance and time criteria for deferral conditions. We need a way to indicate that the unlimited value should be used. I prefer the use of `null` to indicate the unlimited value as opposed to `-1`.

jamesmontemagno commented on Geolocator/Geolocator/Geolocator.Plugin.Abstractions/IGeolocator.cs:
```diff
@@ -67,7 +67,8 @@ public interface IGeolocator
         /// <param name="includeHeading">Include heading or not</param>
-        Task<bool> StartListeningAsync(int minTime, double minDistance, bool includeHeading = false);
+		/// <param name="energySettings">Optional energy settings (iOS only)</param>
+		Task<bool> StartListeningAsync(int minTime, double minDistance, bool includeHeading = false, EnergySettings energySettings = null);
```
Make sure to add the <param comment

MatthewGerber replied:
I'm not sure what you're referring to. There's a `param` comment for the `energySettings` parameter.

jamesmontemagno commented on Geolocator/Geolocator/Geolocator.Plugin.iOS/GeolocatorImplementation.cs:
```diff
@@ -216,8 +218,10 @@ public Task<Position> GetPositionAsync(int timeoutMilliseconds = Timeout.Infinit
-        public Task<bool> StartListeningAsync(int minTime, double minDistance, bool includeHeading = false)
+		public Task<bool> StartListeningAsync(int minTime, double minDistance, bool includeHeading = false, EnergySettings energySettings = null)
         {
+			this.energySettings = energySettings;
```
Right here we should do a check against iOS 6 and if it is < iOS 6 never set it so we never have to check again.

jamesmontemagno commented on Geolocator/Geolocator/Geolocator.Plugin.iOS/GeolocatorImplementation.cs:
```diff
@@ -225,10 +229,23 @@ public Task<bool> StartListeningAsync(int minTime, double minDistance, bool incl
             this.manager.DistanceFilter = minDistance;
-            this.manager.StartUpdatingLocation();
+
+			if (this.energySettings != null && this.energySettings.ListenForSignificantChanges)
```
?. usage would be good, but what happens if DeferLocationUpdate == false ? Then we may run into issues because the above is not set. These should just be one setting with it's default set to true.

MatthewGerber replied:
If `DeferLocationUpdates` is false, then the default desired accuracy and the passed-in minimum distance will be used. This was the previous behavior. Can you clarify what the issue might be?

jamesmontemagno commented:
I left a few more comments in the code. I need to investigate it a bit more an do some refactoring and testing. If you can comment back.

A few concerns I have is that there is a locationManager:didFinishDeferredUpdatesWithError that they say can get called if there is an error. 

https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/LocationAwarenessPG/CoreLocation/CoreLocation.html is a good read through at the very bottom. 

they also have this really odd clause: To prevent itself from calling the allowDeferredLocationUpdatesUntilTraveled:timeout: method multiple times, the delegate uses an internal property to track whether updates have already been deferred. It sets the property to YES in this method and sets it back to NO when deferred updates end.

MatthewGerber commented:
@jamesmontemagno I added a note about the difference between significant change listening and update deferral. Also replied to the other comments.

As for `locationManager:didFinishDeferredUpdatesWithError`, the docs say that if you call `allowDeferredLocationUpdatesUntilTraveled:timeout` multiple times then the previous deferral will be canceled. I'm not sure why this would be bad though. I added a commit to do something similar to the example code.

MatthewGerber commented:
@jamesmontemagno Have you had a chance to inspect the changes up through the following commit:

708f297ab45a22cb649c11758f7e6fcdd61db9ff

If you could release these changes as a beta Nuget I can help test them.

jamesmontemagno commented:
Matthew,

I will be going through a lot of this tonight, it has been a busy time.
Give me 24 hours to pull some stuff down and investigate.

Best,
James

On Wed, Feb 24, 2016 at 1:52 PM, Matthew Gerber <notifications@github.com>
wrote:

> @jamesmontemagno <https://github.com/jamesmontemagno> Have you had a
> chance to inspect the changes up through the following commit:
>
> 708f297
> <https://github.com/jamesmontemagno/Xamarin.Plugins/commit/708f297ab45a22cb649c11758f7e6fcdd61db9ff>
>
> If you could release these changes as a beta Nuget I can help test them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jamesmontemagno/Xamarin.Plugins/pull/226#issuecomment-188468347>
> .
>



-- 
James Montemagno
http://www.montemagno.com
@jamesmontemagno <http://www.twitter.com/jamesmontemagno>
